cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/MacroParser.java

```diff
@@ -0,0 +1,270 @@
+  private static final Pattern COMMA = Pattern.compile(",");
+  private static final int MAX_SUBSTITUTION_DEPTH = 10;
+
+  private MacroParser(){}
+
+  /**
+   * Checks if a string contains a well-formatted macro.
+   * @param str the raw string containing macro syntax.
+   * @throws InvalidMacroException on invalid macro syntax.
+   * @return if str contains a macro.
+   */
+  public static boolean containsMacro(String str) throws InvalidMacroException {
+    // used to check macro syntax
+    substituteMacro(str, null);
+    return (findRightmostMacro(str, null, null) != null);
+  }
+
+  /**
+   * Creates a set of all macros present in a string.
+   * @param str the raw string containing macro syntax.
+   * @throws InvalidMacroException on invalid macro syntax
+   * @return a set of all macros present in str.
+   */
+  public static Set<String> getMacros(String str) throws InvalidMacroException {
+    Set<String> macros = new HashSet<>();
+    substituteMacro(str, 0, null, macros);
+    return macros;
+  }
+
+  /**
+   * Substitutes the provided string with a given macro evaluator. Expands macros from right-to-left recursively.
+   * @param str the raw string containing macro syntax
+   * @param macroEvaluator the evaluator used to substitute macros
+   * @return the original string with all macros expanded and substituted
+   */
+  public static String substituteMacro(String str, MacroEvaluator macroEvaluator) {
+    // final string has escapes not directly embedded in macro syntax replaced
+    return replaceEscapedSyntax(substituteMacro(str, 0, macroEvaluator, null));
+  }
+
+  private static String substituteMacro(String str, int depth, @Nullable MacroEvaluator macroEvaluator,
+                                        @Nullable Set<String> macros) {
+    if (depth > MAX_SUBSTITUTION_DEPTH) {
+      throw new InvalidMacroException(String.format("Failed substituting maco '%s', expansion exceeded %d levels.",
+                                                    str, MAX_SUBSTITUTION_DEPTH));
+    }
+
+    MacroMetadata macroPosition = findRightmostMacro(str, macroEvaluator, macros);
+    while (macroPosition != null) {
+      try {
+        str = str.substring(0, macroPosition.startIndex) +
+          substituteMacro(macroPosition.substitution, depth + 1, macroEvaluator, macros) +
+          str.substring(macroPosition.endIndex + 1);
+      } catch (InvalidMacroException e) {
+        throw new InvalidMacroException(e.getMessage());
+      }
+      macroPosition = findRightmostMacro(str, macroEvaluator, macros);
+    }
+    return str;
+  }
+
+  /**
+   * Find the rightmost macro in the specified string. If no macro is found, returns null.
+   *
+   * @param str the string to find a macro in
+   * @return the rightmost macro and its position in the string
+   * @throws InvalidMacroException if invalid macro syntax was found. This cannot be thrown if isLenient is true.
```

Comment:
@awholegunch means you have something in javadoc which is not there in the parameters. so remove the javadoc.